src/toast/pipeline_tools/madam.py

```diff
@@ -0,0 +1,461 @@
+
+    Bin and optionally destripe all conceivable subsets of the data.
+
+    Args:
+        freq (str) :  Frequency identifier to append to the file prefix
+        time_comms (iterable) :  Series of disjoint communicators that
+            map, e.g., seasons and days.  Each entry is a tuple of
+            the form (`name`, `communicator`)
+        telescope_data (iterable) : series of disjoint TOAST data
+            objects.  Each entry is tuple of the form (`name`, `data`).
+        bin_only (bool) :  Disable destriping and only bin the signal,
+            Useful when running Madam as a part of a filter+bin pipeline.
+    """
+    if comm.comm_world is None:
+        raise RuntimeError("Madam requires MPI")
+
+    log = Logger.get()
+    timer = Timer()
+    timer.start()
+    if comm.world_rank == 0 and verbose:
+        log.info("Making maps")
+
+    pars = copy.deepcopy(madampars)
+    pars["path_output"] = outpath
+    if comm.world_rank == 0:
+        os.makedirs(outpath, exist_ok=True)
+    file_root = pars["file_root"]
+    if extra_prefix is not None:
+        if len(file_root) > 0 and not file_root.endswith("_"):
+            file_root += "_"
+        file_root += "{}".format(extra_prefix)
+    if freq is not None:
+        if len(file_root) > 0 and not file_root.endswith("_"):
+            file_root += "_"
+        file_root += "{:03}".format(int(freq))
+
+    if not first_call:
+        # Only the first MC iteration should produce the hits and
+        # white noise matrices
+        pars["write_matrix"] = False
+        pars["write_wcov"] = False
+        pars["write_hits"] = False
+
+    if bin_only:
+        pars["kfirst"] = False
+        pars["write_map"] = False
+        pars["write_binmap"] = True
+
+    # Sanity check, is any of the Madam outputs required?
+
+    outputs = [
+        pars["write_map"],
+        pars["write_binmap"],
+        pars["write_hits"],
+        pars["write_wcov"],
+        pars["write_matrix"],
+    ]
+    if not np.any(outputs):
+        if comm.world_rank == 0:
+            log.info("No Madam outputs requested.  Skipping.")
+        return
+
+    if args.madam_noisefilter or not pars["kfirst"]:
+        # With the noise filter enabled, we want to enforce continuity
+        # across the Observation.  Otherwise we fit each interval
+        # separately.
+        madam_intervals = None
+    else:
+        madam_intervals = "intervals"
+
+    madam = OpMadam(
+        params=pars,
+        detweights=detweights,
+        name=cache_name,
+        common_flag_mask=args.common_flag_mask,
+        purge_tod=False,
+        intervals=madam_intervals,
+        conserve_memory=args.conserve_memory,
+    )
+
+    if "info" in madam.params:
+        info = madam.params["info"]
+    else:
+        info = 3
+
+    if time_comms is None:
+        time_comms = [("all", comm.comm_world)]
+
+    if telescope_data is None:
+        telescope_data = [("all", data)]
+
+    ttimer = Timer()
```

Comment:
what does `ttimer` stands for? maybe add only comment if variable name gets too long

Reply:
Ted likes very short variable names. I replaced `timer` -> `total_timer` and `ttimer` with `timer`.